### PR TITLE
GEOMESA-1586 Implement Geometric Constructor SQL Functions

### PR DESCRIPTION
**Pull request by cfkelly**
Signed-off-by: Charles Kelly <charles.kelly@ccri.com>

### CODE REVIEW

File: geomesa-spark/geomesa-spark-sql/src/main/scala/org/apache/spark/sql/SQLCastFunctions.scala [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+package org.apache.spark.sql
+
+import com.vividsolutions.jts.geom._
+import org.apache.spark.sql.SQLContext
+
+object SQLCastFunctions {
+  val ST_CastToPoint:      Geometry => Point       = g => g.asInstanceOf[Point]
+  val ST_CastToPolygon:    Geometry => Polygon     = g => g.asInstanceOf[Polygon]
+  val ST_CastToLineString: Geometry => LineString  = g => g.asInstanceOf[LineString]
+  val ST_ByteArray: (String) => Array[Byte] = (string) => string.getBytes
```
Comment (elahrvivaz) [inline geomesa-spark/geomesa-spark-sql/src/main/scala/org/apache/spark/sql/SQLCastFunctions.scala]: we should probably use `.getBytes(StandardCharsets.UTF_8)`

File: geomesa-spark/geomesa-spark-sql/src/main/scala/org/apache/spark/sql/SQLConstructorFunctions.scala [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+package org.apache.spark.sql
+
+import com.vividsolutions.jts.geom._
+import org.geotools.geometry.jts.JTS
+import org.locationtech.geomesa.utils.geohash.{BoundingBox, GeoHash}
+import org.locationtech.geomesa.utils.text.{WKBUtils, WKTUtils}
+import org.apache.spark.sql.SQLContext
+
+object SQLConstructorFunctions {
```
Comment (mdzimmerman) [inline geomesa-spark/geomesa-spark-sql/src/main/scala/org/apache/spark/sql/SQLConstructorFunctions.scala]: Can we rename to ``SQLGeometricConstructorFunctions``?

File: geomesa-spark/geomesa-spark-sql/src/main/scala/org/apache/spark/sql/SQLCastFunctions.scala [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+* which accompanies this distribution and is available at
+* http://www.opensource.org/licenses/apache2.0.php.
+*************************************************************************/
+
+package org.apache.spark.sql
+
+import com.vividsolutions.jts.geom._
+import org.apache.spark.sql.SQLContext
+
+object SQLCastFunctions {
```
Comment (mdzimmerman) [inline geomesa-spark/geomesa-spark-sql/src/main/scala/org/apache/spark/sql/SQLCastFunctions.scala]: Rename to ``SQLGeometricCastFunctions``?

File: geomesa-spark/geomesa-spark-sql/src/main/scala/org/apache/spark/sql/SQLTypes.scala [hunk 4]
```diff
@@ -41,7 +41,9 @@ object SQLTypes {
   UDTRegistration.register(classOf[Geometry].getCanonicalName, classOf[GeometryUDT].getCanonicalName)
 
   def init(sqlContext: SQLContext): Unit = {
+    SQLCastFunctions.registerFunctions(sqlContext)
```
Comment (mdzimmerman) [inline geomesa-spark/geomesa-spark-sql/src/main/scala/org/apache/spark/sql/SQLTypes.scala]: +1 for splitting these into different classes.

File: geomesa-spark/geomesa-spark-sql/src/test/scala/org/locationtech/geomesa/spark/SparkSQLGeometryConstructorsTest.scala [hunk 5]
```diff
@@ -53,14 +55,72 @@ class SparkSQLGeometryConstructorsTest extends Specification with LazyLogging {
       df.collect().length mustEqual 3
     }
 
+    "st_box2DFromGeoHash" >> {
+      val r = sc.sql(
+        s"""
+           |select st_box2DFromGeoHash('u2sux', 25)
```
Comment (mdzimmerman) [inline geomesa-spark/geomesa-spark-sql/src/test/scala/org/locationtech/geomesa/spark/SparkSQLGeometryConstructorsTest.scala]: Perhaps we should consider adding an example that can be verified independently; such as the "ezs42" example  described on the geohash wikipedia page:  https://en.wikipedia.org/wiki/Geohash

File: geomesa-spark/geomesa-spark-sql/src/main/scala/org/apache/spark/sql/SQLConstructorFunctions.scala [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+  val ST_GeomFromWKT: String => Geometry = text => WKTUtils.read(text)
+  val ST_GeomFromWKB: Array[Byte] => Geometry = array => WKBUtils.read(array)
+  val ST_MakeBox2D: (Point, Point) => Polygon = (ll, ur) =>
+    JTS.toGeometry(new Envelope(ll.getX, ur.getX, ll.getY, ur.getY))
+  val ST_MakeBBOX: (Double, Double, Double, Double) => Geometry = (lx, ly, ux, uy) =>
+    JTS.toGeometry(BoundingBox(lx, ux, ly, uy))
+  val ST_MakePolygon: LineString => Polygon = shell => {
+    require(shell.isClosed)
+    val envelope = JTS.toEnvelope(new LinearRing(shell.getCoordinateSequence, SQLTypes.geomFactory).getEnvelope)
+    JTS.toGeometry(envelope)
```
Comment (cfkelly) [inline geomesa-spark/geomesa-spark-sql/src/main/scala/org/apache/spark/sql/SQLConstructorFunctions.scala]: good catch!

**Comment by jnh5y** [comment]
Ok, after today's discussion, it seems like we could have a GeometryConstructorRule.  Should that be in this PR or a follow-on one?

File: geomesa-spark/geomesa-spark-sql/src/main/scala/org/apache/spark/sql/SQLGeometricConstructorFunctions.scala [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    sqlContext.udf.register("st_geomFromWKB"       , ST_GeomFromWKB)
+    sqlContext.udf.register("st_makeBox2D"         , ST_MakeBox2D)
+    sqlContext.udf.register("st_makeBBOX"          , ST_MakeBBOX)
+    sqlContext.udf.register("st_makePolygon"       , ST_MakePolygon)
+    sqlContext.udf.register("st_makePoint"         , ST_MakePoint)
+    sqlContext.udf.register("st_makePointM"        , ST_MakePointM)
+    sqlContext.udf.register("st_mLineFromText"     , ST_MLineFromText)
+    sqlContext.udf.register("st_mPointFromText"    , ST_MPointFromText)
+    sqlContext.udf.register("st_mPolyFromText"     , ST_MPolyFromText)
+    sqlContext.udf.register("ST_point"             , ST_Point)
```
Comment (Afterglow375) [inline geomesa-spark/geomesa-spark-sql/src/main/scala/org/apache/spark/sql/SQLGeometricConstructorFunctions.scala]: lowercase the ST?

File: geomesa-spark/geomesa-spark-sql/src/main/scala/org/apache/spark/sql/SQLGeometricConstructorFunctions.scala [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+
+package org.apache.spark.sql
+
+import com.vividsolutions.jts.geom._
+import org.geotools.geometry.jts.JTS
+import org.locationtech.geomesa.utils.geohash.{BoundingBox, GeoHash}
+import org.locationtech.geomesa.utils.text.{WKBUtils, WKTUtils}
+import org.apache.spark.sql.SQLContext
+
+object SQLGeometricConstructorFunctions {
```
Comment (Afterglow375) [inline geomesa-spark/geomesa-spark-sql/src/main/scala/org/apache/spark/sql/SQLGeometricConstructorFunctions.scala]: Wondering how these functions should be documented. Relying on the PostGIS documentation doesn't seem good as there are probably slight differences.

File: geomesa-spark/geomesa-spark-sql/src/main/scala/org/apache/spark/sql/SQLTypes.scala [hunk 9]
```diff
@@ -41,7 +41,9 @@ object SQLTypes {
   UDTRegistration.register(classOf[Geometry].getCanonicalName, classOf[GeometryUDT].getCanonicalName)
 
   def init(sqlContext: SQLContext): Unit = {
+    SQLGeometricCastFunctions.registerFunctions(sqlContext)
```
Comment (Afterglow375) [inline geomesa-spark/geomesa-spark-sql/src/main/scala/org/apache/spark/sql/SQLTypes.scala]: I personally put my registerFunction call within the `SQLSpatialFunctions.registerFunctions(sqlContext)` function. Don't which approach makes the most sense.

File: geomesa-spark/geomesa-spark-sql/src/main/scala/org/apache/spark/sql/SQLGeometricConstructorFunctions.scala [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+  val ST_GeomFromGeoHash: (String, Int) => Geometry = (hash, prec) => GeoHash(hash, prec).geom
+  val ST_Box2DFromGeoHash: (String, Int) => Geometry = (hash, prec) => ST_GeomFromGeoHash(hash, prec)
+  val ST_GeomFromWKT: String => Geometry = text => WKTUtils.read(text)
+  val ST_GeomFromWKB: Array[Byte] => Geometry = array => WKBUtils.read(array)
+  val ST_MakeBox2D: (Point, Point) => Polygon = (ll, ur) =>
+    JTS.toGeometry(new Envelope(ll.getX, ur.getX, ll.getY, ur.getY))
+  val ST_MakeBBOX: (Double, Double, Double, Double) => Geometry = (lx, ly, ux, uy) =>
```
Comment (Afterglow375) [inline geomesa-spark/geomesa-spark-sql/src/main/scala/org/apache/spark/sql/SQLGeometricConstructorFunctions.scala]: [picky] these variable names are obscure.

File: geomesa-spark/geomesa-spark-sql/src/main/scala/org/apache/spark/sql/SQLGeometricConstructorFunctions.scala (hunk 8 above)
Comment (jnh5y) [inline geomesa-spark/geomesa-spark-sql/src/main/scala/org/apache/spark/sql/SQLGeometricConstructorFunctions.scala]: @Afterglow375 I agree that we ought not link to PostGIS's documentation. 

@mdzimmerman ideas for documenting the functions easily?

File: geomesa-spark/geomesa-spark-sql/src/main/scala/org/apache/spark/sql/SQLTypes.scala (hunk 9 above)
Comment (jnh5y) [inline geomesa-spark/geomesa-spark-sql/src/main/scala/org/apache/spark/sql/SQLTypes.scala]: @Afterglow375, let's huddle up sometime today and figure out which approach we prefer.

**Pull request merged**